### PR TITLE
Support for sample filtering by thread

### LINKED ISSUE #70: Add filter by thread
Currently, when multiple threads are selected for a profiling session, stats are displayed as an aggregate for all threads. When trying to profile a task system for example, you don't know in advance which thread will pick up the workload you're interested in profiling, so you can't select it beforehand.
The ability to filter by thread would alleviate the issue.

### PR DESCRIPTION
**Pull request by aaalexandrov**
Implement source thread-aware sample gathering, file format and display in the UI.
Add sample filtering by thread.
Add symbol per-thread statistics UI.
Fixes #70

### CODE REVIEW

**Comment by CyberShadow** [comment]
I can't open old `.sleepy` files with this change, could you please check?
We should support loading old files even when the format for new files changes.

**Comment by aaalexandrov** [comment]
I did not add compatibility code for opening the old file format files because I saw no compatibility code there for any old versions.
There didn't seem to be any infrastructure for that in place, aside from comparing to the current version and aborting in case it's different.
I can add loading from 0.90 specifically, but not backwards from there.
Out of curiosity, how often does this format change?

**Comment by CyberShadow** [comment]
> I did not add compatibility code for opening the old file format files because I saw no compatibility code there for any old versions.

You're right, my mistake. I thought I remembered some compatibility code, but I guess I was wrong. So, I take that back (though of course, speaking of a separate enhancement, compatibility would be nice if possible).

>  Out of curiosity, how often does this format change?

Well, the 0.90 version was bumped in 2014...

**Comment by aaalexandrov** [comment]
I'll do compatibility from 0.90 in the coming days, it should be straightforward enough.

**Comment by CyberShadow** [comment]
OK, only if you want (my initial statement was based on a bad assumption, and practically speaking, it's easy enough to download an older version to open older files).

Since you're working with threads, maybe looking at #4 would be a better use of your time.

As for this PR, I will try to have a closer look at it and merge it soon.

**Comment by aaalexandrov** [comment]
I'll see what I can do about sampling newly spawned threads.
Is there a better way to detect them than just polling and comparing to the threads we're already tracking?

**Comment by CyberShadow** [comment]
I believe that a long time ago, @rmitton tried to implement this by making Very Sleepy also work as a debugger, attaching itself to the process. That would allow it to be notified of new threads using `CREATE_PROCESS_DEBUG_EVENT`.

There is code for this in [src/profiler/debugger.cpp](https://github.com/verysleepy/verysleepy/blob/master/src/profiler/debugger.cpp), but it was never finished.

See also #20.

File: src/profiler/profilerthread.cpp [hunk 1]
```diff
@@ -40,15 +41,19 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 
 // DE: 20090325: Profiler has a list of threads to profile
 // RM: 20130614: Profiler time can now be limited (-1 = until cancelled)
-ProfilerThread::ProfilerThread(HANDLE target_process_, const std::vector<HANDLE>& target_threads, SymbolInfo *sym_info_)
+ProfilerThread::ProfilerThread(HANDLE target_process_, const std::vector<std::pair<HANDLE, DWORD>>& target_threads, SymbolInfo *sym_info_)
```
Comment (CyberShadow) [inline src/profiler/profilerthread.cpp:44]: Would it be simpler to use `GetThreadId` instead of additionally passing the ID around?

File: src/profiler/profiler.h [hunk 2]
```diff
@@ -105,8 +114,7 @@ class Profiler
 	~Profiler();
 
 	// DE: 20090325: Profiler no longer owns callstack and flatcounts since it is shared between multipler profilers
-	std::map<CallStack, SAMPLE_TYPE>& callstacks;
-	std::map<PROFILER_ADDR, SAMPLE_TYPE>& flatcounts;
```
Comment (CyberShadow) [inline src/profiler/profiler.h:109]: It would be nice if you could add a description (a few paragraphs) of the changes in this PR in the commit message. For example, it could mention why `flatcounts` were removed, and what replaces them (if anything). This will help future contributors as well.

File: src/profiler/symbolinfo.cpp [hunk 3]
```diff
@@ -158,9 +158,9 @@ void SymbolInfo::loadSymbolsUsing(DbgHelp* dbgHelp, const std::wstring& sympath)
 			// This is a secondary dbgHelp, so just complement debug
 			// information for modules that have none.
 			
-			for (size_t n=0;n<modules.size();n++)
+			for (size_t m=0;m<modules.size();m++)
 			{
-				Module &mod = modules[n];
+				Module &mod = modules[m];
```
Comment (CyberShadow) [inline src/profiler/symbolinfo.cpp:163]: It might be too late for that, but generally it's better to place unrelated changes in separate commits. That would also provide for a place to describe why the change is done (i.e. which compiler warning is fixed if any), in the commit message.

File: src/profiler/threadinfo.cpp [hunk 4]
```diff
@@ -36,34 +36,41 @@ static __int64 getTotal(FILETIME time)
...
 {
 	// Helper function to let main window decide whether to hide this column
-	return GetThreadDescription != NULL;
+	return GetThreadDescription_ != NULL;
 }
 
-// DE: 20090325 Threads now have CPU usage
-ThreadInfo::ThreadInfo(DWORD id_, HANDLE thread_handle_)
-:	id(id_), thread_handle(thread_handle_)
+std::wstring getThreadDescriptorName(HANDLE thread_handle)
```
Comment (CyberShadow) [inline src/profiler/threadinfo.cpp:47]: Refactoring this function out also would ideally be done in its own commit.

File: src/wxProfilerGUI/CallstackView.cpp [hunk 5]
```diff
@@ -153,7 +148,9 @@ void CallstackView::showCallStack(const Database::Symbol *symbol)
 	}
 
 	callstacks = database->getCallstacksContaining(symbol);
-	std::sort(callstacks.begin(),callstacks.end(),SortCalls);
+	std::sort(callstacks.begin(), callstacks.end(), [this](const Database::CallStack *a, const Database::CallStack *b) {
+		return database->getFilteredSampleCount(a->samples) > database->getFilteredSampleCount(b->samples);
+	});
```
Comment (CyberShadow) [inline src/wxProfilerGUI/CallstackView.cpp:153]: :warning: I believe this change increases the complexity of the sort from O(n log n) to O(n^2 (log n) (log n)). I'm concerned that this may become a problem on larger profiling results.

File: src/wxProfilerGUI/database.cpp [hunk 6]
```diff
@@ -243,6 +243,45 @@ void Database::loadSymbols(wxInputStream &file)
 	progressdlg.Update(kMaxProgress, "Tidying things up...");
 }
 
+static void addSampleInfo(std::vector<Database::SampleInfo> &samples, Database::SampleInfo const &info)
+{
+	auto it = std::lower_bound(samples.begin(), samples.end(), info, [](Database::SampleInfo const &i0, Database::SampleInfo const &i1) { return i0.threadid < i1.threadid; });
+	if (it == samples.end() || it->threadid != info.threadid) {
+		samples.insert(it, info);
```
Comment (CyberShadow) [inline src/wxProfilerGUI/database.cpp:250]: Inserting into a vector is an O(n) operation. I'm not sure how many of those we've had before, but if this is new, it may result in a performance regression for longer profiling sessions.

**Review by CyberShadow — COMMENTED** [review]
Here are my observations so far. I am going to try and see if this results in any changes in performance in practice.

File: src/wxProfilerGUI/database.cpp (hunk 6 above)
Comment (aaalexandrov) [reply src/wxProfilerGUI/database.cpp:250]: New entries are only inserted if a sample is taken from the context of a new thread. If the profiled application starts and retires an unbounded number of threads, and we encounter callstack samples called from those threads, then yes - inserting might become an issue. Currently the profiled thread number is constant however. A potential solution to improve the O() complexity would be to use a map instead of a vector. Given the implementation of a map as a tree is not cache-friendly, I chose to use an array since that would be faster for the current/typical use case. If you're more concerned with the worst, rather than the typical case performance however, I can change the array to a map?

File: src/profiler/symbolinfo.cpp (hunk 3 above)
Comment (aaalexandrov) [reply src/profiler/symbolinfo.cpp:163]: I assumed you'd prefer a single commit. I squashed what few separate commits I had into one before pushing, sorry about that.
The local variable name changes here and there are to fix class / outer scope variables being shadowed.

File: src/wxProfilerGUI/CallstackView.cpp (hunk 5 above)
Comment (CyberShadow) [reply src/wxProfilerGUI/CallstackView.cpp:153]: Would it make sense to do a Schwartzian transform here?

File: src/wxProfilerGUI/database.cpp (hunk 6 above)
Comment (CyberShadow) [reply src/wxProfilerGUI/database.cpp:250]: Thanks for the explanation. You're probably right, we (or the user otherwise) will probably start running into performance problems elsewhere before this becomes an issue.

File: src/profiler/symbolinfo.cpp (hunk 3 above)
Comment (CyberShadow) [reply src/profiler/symbolinfo.cpp:163]: Some good advice can be found by asking the right question:

https://www.google.com/search?q=when+to+commit+git

For instance:

> You shouldn't commit based on a time basis, but on a feature basis. Whenever you add a new feature that's worth committing, commit. You added a working method? Commit. You fixed a typo? Commit. You fixed a file's wrong indentation? Commit. There's nothing wrong committing small changes, as soon as the commit is relevant.
> 
> What is wrong is committing a huge number of changes, with no relations between each others. It makes it very hard to detect the commit source of a given regression.


-- https://softwareengineering.stackexchange.com/a/74893/16481

File: src/wxProfilerGUI/database.cpp [hunk 7]
```diff
@@ -255,27 +294,53 @@ void Database::loadCallstacks(wxInputStream &file,bool collapseKernelCalls)
...
+			streamAddr >> addrstr;
 			if (addrstr.empty())
 				break;
 			Address addr = hexStringTo64UInt(addrstr);
 
+			if (depth == 0)
+			{
+				topAddrInfo = &addrinfo.at(addr);
+			}
+
```
Comment (CyberShadow) [inline src/wxProfilerGUI/database.cpp:320]: So I guess this is `IPCounts.txt`'s replacement.

File: src/wxProfilerGUI/proclist.cpp [hunk 8]
```diff
@@ -132,8 +132,16 @@ void ProcList::OnContextMenu(wxContextMenuEvent& WXUNUSED(event))
...
+struct ExclusivePred  { bool operator () (const Database::Item &a, const Database::Item &b) {
+	if (a.exclusive != b.exclusive)
+		return a.exclusive < b.exclusive;
+	return a.inclusive < b.inclusive;
+} };
+struct InclusivePred  { bool operator () (const Database::Item &a, const Database::Item &b) {
+	if (a.inclusive != b.inclusive)
+		return a.inclusive < b.inclusive;
+	return a.exclusive < b.exclusive;
+} };
```
Comment (CyberShadow) [inline src/wxProfilerGUI/proclist.cpp:144]: Nice, but ideally should be in its own commit. As it is there is zero documentation about this improvement.

File: src/wxProfilerGUI/threadsview.cpp [hunk 9]
```diff
@@ -0,0 +1,387 @@
+/*=====================================================================
+threadsview.cpp
+------------
+
+This program is free software; you can redistribute it and/or
```
Comment (CyberShadow) [inline src/wxProfilerGUI/threadsview.cpp:5]: The license header is void without a copyright line.

If you don't want to add your name, we can write something like "Copyright (C) Very Sleepy authors and contributors".

File: src/wxProfilerGUI/CallstackView.cpp (hunk 5 above)
Comment (aaalexandrov) [reply src/wxProfilerGUI/CallstackView.cpp]: I did that, it should address any potential slowdown there.

File: src/wxProfilerGUI/database.cpp (hunk 6 above)
Comment (aaalexandrov) [reply src/wxProfilerGUI/database.cpp]: Changed the storage to a map instead of sorted array, it should fix the potential worst case behavior while the typical case slowdown should not be noticeable to the user.

File: src/wxProfilerGUI/database.cpp (hunk 7 above)
Comment (aaalexandrov) [reply src/wxProfilerGUI/database.cpp]: That's right. Since the sample count is no longer a single number per address, I figured it's better to reconstruct it at load time rather than record/write/read it twice.

**Comment by CyberShadow** [comment]
Great! Thank you, and I don't have any more suggestions :)

Since I have a good setup for it, I am going to try and massage the git history a bit to ungroup unrelated changes from the big commits and squash changes that are not relevant to the history.

**Comment by aaalexandrov** [comment]
Thanks, if there's anything I missed, don't hesitate to say.
I'll probably do the issue about profiling newly spawned threads at some point, but it'll take me some time.

File: src/wxProfilerGUI/profilergui.cpp [hunk 10]
```diff
@@ -337,9 +337,12 @@ static std::vector<HANDLE> getThreadsByAttachMode(ProcessInfo& process_info)
 		return threadHandles;
 
 	case ATTACH_MOST_BUSY_THREAD:
-		if (HANDLE mostBusy = getMostBusyThread(process_info))
+	{
+		HANDLE mostBusy = getMostBusyThread(process_info);
+		if (mostBusy)
 			threadHandles.push_back(mostBusy);
 		return threadHandles;
+	}
```
Comment (CyberShadow) [inline src/wxProfilerGUI/profilergui.cpp:345]: There might be a better way to silence this warning.

Depending on the compiler, sometimes wrapping the entire `if` expression in another set of parens does it.

**Review by CyberShadow — APPROVED** [review]
Let me know if you agree with my history refactoring and commit messages, and we can merge it.

**Comment by aaalexandrov** [comment]
It looks great to me, thanks for separating the commits.

**Comment by CyberShadow** [comment]
Thanks for the very nice contribution :)
